diff --git a/mrbgems/picoruby-ble/ports/rp2040/ble.c b/mrbgems/picoruby-ble/ports/rp2040/ble.c
index d0a28e50..41178b12 100644
--- a/mrbgems/picoruby-ble/ports/rp2040/ble.c
+++ b/mrbgems/picoruby-ble/ports/rp2040/ble.c
@@ -18,6 +18,7 @@ static gatt_client_notification_t notification_listener;
 static gatt_client_characteristic_t notification_characteristic;
 static gatt_client_notification_t all_notification_listener;
 static bool all_notification_listener_registered = false;
+static bool central_hci_working = false;

 static uint16_t
 att_read_callback(hci_con_handle_t connection_handle, uint16_t att_handle, uint16_t offset, uint8_t *buffer, uint16_t buffer_size)
@@ -78,7 +79,18 @@ packet_handler(uint8_t packet_type, uint16_t channel, uint8_t *packet, uint16_t
     case BLE_ROLE_CENTRAL:
       switch (hci_event_packet_get_type(packet)) {
         case BTSTACK_EVENT_STATE:
+          central_hci_working = (btstack_event_state_get_state(packet) == HCI_STATE_WORKING);
+          BLE_push_event(packet, size);
+          break;
+        case HCI_EVENT_COMMAND_COMPLETE:
+        case HCI_EVENT_COMMAND_STATUS:
+          if (central_hci_working) {
+            BLE_push_event(packet, size);
+          }
+          break;
+        case HCI_EVENT_DISCONNECTION_COMPLETE:
         case HCI_EVENT_LE_META:
+        case HCI_EVENT_META_GAP:
         case GAP_EVENT_ADVERTISING_REPORT:
         case GATT_EVENT_QUERY_COMPLETE:
         case GATT_EVENT_SERVICE_QUERY_RESULT:
         case GATT_EVENT_CHARACTERISTIC_QUERY_RESULT:
