diff --git a/mrbgems/picoruby-ble/include/btstack_config.h b/mrbgems/picoruby-ble/include/btstack_config.h
--- a/mrbgems/picoruby-ble/include/btstack_config.h
+++ b/mrbgems/picoruby-ble/include/btstack_config.h
@@ -15,5 +15,5 @@
 // for the client
 #define ENABLE_LE_CENTRAL
-#define MAX_NR_GATT_CLIENTS 1
+#define MAX_NR_GATT_CLIENTS 2

 // BTstack configuration. buffers, sizes, ...
@@ -21,5 +21,5 @@
 #define HCI_ACL_PAYLOAD_SIZE (255 + 4)
 #define HCI_ACL_CHUNK_SIZE_ALIGNMENT 4
-#define MAX_NR_HCI_CONNECTIONS 1
+#define MAX_NR_HCI_CONNECTIONS 2
 #define MAX_NR_SM_LOOKUP_ENTRIES 3
 #define MAX_NR_WHITELIST_ENTRIES 16
