diff --git a/mrbgems/picoruby-ble/src/mruby/ble.c b/mrbgems/picoruby-ble/src/mruby/ble.c
index ce314f54..0d40addd 100644
--- a/mrbgems/picoruby-ble/src/mruby/ble.c
+++ b/mrbgems/picoruby-ble/src/mruby/ble.c
@@ -8,11 +8,24 @@ static mrb_state *_mrb = NULL;
 static mrb_value write_values;
 static mrb_value read_values;
 
+#define BLE_EVENT_BTSTACK_STATE 0x60
+
 void
 BLE_push_event(uint8_t *data, uint16_t size)
 {
   if (packet_mutex) return;
   packet_mutex = true;
+  if (
+    packet_flag &&
+    packet != NULL &&
+    0 < packet_size &&
+    0 < size &&
+    packet[0] == BLE_EVENT_BTSTACK_STATE &&
+    data[0] != BLE_EVENT_BTSTACK_STATE
+  ) {
+    packet_mutex = false;
+    return;
+  }
   packet_flag = true;
   packet_size = size;
   if (packet != NULL) {
