diff --git a/mrbgems/picoruby-ble/src/mruby/ble_central.c b/mrbgems/picoruby-ble/src/mruby/ble_central.c
index 1fe39696..149160e8 100644
--- a/mrbgems/picoruby-ble/src/mruby/ble_central.c
+++ b/mrbgems/picoruby-ble/src/mruby/ble_central.c
@@ -103,7 +103,7 @@ mrb_set_scan_params(mrb_state *mrb, mrb_value self)
     scan_type_num = 1;
     mrb_notimplement(mrb); // TODO
   } else if (mrb_symbol(scan_type) == MRB_SYM(passive)) {
-    scan_type_num = 1;
+    scan_type_num = 0;
   } else {
     mrb_raise(mrb, E_ARGUMENT_ERROR, "wrong type of scan_type");
   }
